src/analysis/cache.Rmd

```diff
@@ -0,0 +1,313 @@
+
+### LLC_loads, LLC_load_misses
+
+The miss ratio for the last level cache is very low, 0.02% at most. See the table below.
+
+```{r}
+sqldf("SELECT opcode, max(LLC_load_misses/LLC_loads) as 'max LLC miss ratio' FROM measurements_avg GROUP BY opcode")
+```
+
+One observation is that our tests hardly leave caches. The other observation is that every opcode has almost the same ratio, which is very good.
+
+Finally, we compare usage of L1 caches and LLC caches. The formula is `LLC_loads/(L1_icache_loads+L1_dcache_loads)`
+
+```{r}
+sqldf("SELECT opcode, avg(LLC_loads/(L1_icache_loads+L1_dcache_loads)) as ratio, stdev(LLC_loads/(L1_icache_loads+L1_dcache_loads)) as sdev FROM measurements_avg GROUP BY opcode")
+```
+
+Every opcode has similar ratio. L1 cache is 1000 times more often in use than LLC. So L1 cache performance has crucial impact on test results.
+
+### page_faults
+
+The graph of page_faults is saw shaped. But the absolute values are much higher so changes are relatively low.
+We calculate the range of page_faults for each op_count (max-min).
+
+```{r}
+sqldf("SELECT op_count, max(page_faults)-min(page_faults) as 'range (max-min)' FROM measurements_avg GROUP BY op_count")
+```
+
+These ranges are equal 3 at most which is around 1% of the average values. There is the saw effect. But the amplitudes of opcodes are circa 6 which is around 2.5% of the average values. So we say that page_faults does not depend on opcode actually and almost do not depend on op_counts. The presumption is that page_faults are due to startup/warmup, but this requires further analysis.
+
+```{r}
+sqldf("SELECT opcode, max(page_faults)-min(page_faults) as 'amplitude (max-min)' FROM measurements_avg GROUP BY opcode")
+```
+
+### branches, branch_misses
+
+Correlation of branches is more than 0.99 for every opcode. This is not surprising. But the distribution is different than task_clock, if we consider slopes. And that is not clear why.
+
+Here are the miss ratios for the branch prediction. 
+
+```{r}
+sqldf("SELECT opcode, avg(branch_misses/branches) as 'miss ratio', stdev(branch_misses/branches) as sdev FROM measurements_avg GROUP BY opcode")
+```
+
+Note that the branch prediction is very effective for every opcode, the hit ratio is around 0.999. And the fact that branch_misses are not linear does not impact these ratios much. So we can say that branch_misses does not impact on opcode usage cost because of low values, and branches are safe factors because of high correlation.
+
+Lets look closer at branch_misses graph. There are two steps. At op_count=0 the values are 5.2M. Then there is a jump to 6.2M at op_count=5. Not all opcodes behaves the same way, but the deviations are not significant. The general conclusion is that there is a visible increase of branch_misses when a new opcode is added to a program, but this increase does not depend on an opcode itself and does not depend on the number of opcode occurrences.
```

Comment:
> Not all opcodes behaves the same way

I can see all opcodes have the step at op_count=5, what do you mean by "behave the same way"? to me they do.